resourceIndexer/inventory.js

```diff
@@ -0,0 +1,255 @@
+          `.replace(/\s+/g, ' '))
+        }
+      }
+      return item;
+    })
+}
+
+/**
+ * For a non-versioned object, check that the VersionId is _not_ present and that keys are never duplicated.
+ * Adds the property 'IsLatest' = 'true' for every object
+ */
+function _checkNonVersionedObjects(objects) {
+  const keys = new Set();
+  objects.forEach((item) => {
+    if (item.hasOwnProperty('VersionId')) { // eslint-disable-line no-prototype-builtins
+      logger.verbose(`Object ${item.Bucket}/${item.Key} has a versionId ('${item.VersionId}') but the bucket is not versioned! The item will be ignored.`);
+      return false;
+    }
+    if (keys.has(item.Key)) {
+      throw new ResourceIndexerError(`
+        The S3 Object for ${item.Bucket}/${item.Key} (unexpectedly) appears multiple times in an un-versioned bucket.
+        This may result in a corrupted index and so is a fatal error.
+      `.replace(/\s+/g, ' '))
+    }
+    keys.add(item.Key);
+  })
+  return objects;
+
+}
+
+/**
+ * Removed 'deleted' objects in a (versioned) bucket. Versions more recent than
+ * the most recent delete marker will _not_ be removed. Delete markers
+ * themselves will be removed.
+ *
+ * Non-versioned buckets don't have delete markers, and it's safe to run this
+ * function for them.
+ */
+function _removeDeletedObjects(objects) {
+
+  // Store the most recent delete makers. Keys are s3 keys, values are the timestamp of the delete marker
+  const deleteMarkers = {};
+
+  return objects
+    .filter((item) => {
+      if (item.IsDeleteMarker === "true") {
+        if (!deleteMarkers[item.Key] || deleteMarkers[item.Key]<item.timestamp) {
+          deleteMarkers[item.Key] = item.timestamp;
+        }
+        return false;
+      }
+      return true;
+    })
+    .filter((item) => {
+      if (deleteMarkers[item.Key]) {
+        if (item.timestamp <= deleteMarkers[item.Key]) {
+          return false;
+        }
+      }
+      return true;
+    })
+}
+
+/**
+ * Parses a S3 inventory manifest JSON file
+ * @param {object} manifest
+ * @returns {object} object.schema = string[]
+ *                   object.inventoryKey = string
+*/
+function _parseManifest(manifest) {
+  if (manifest.files.length>1) {
+    throw new ResourceIndexerError(`
+      The manifest file for the S3 inventory for bucket ${manifest.sourceBucket}
+      includes more than one inventory file. This situation was not encountered
+      during development, but this is presumably caused by the inventory size
+      exceeding some threshold and being chunked into multiple files. Please check
+      this is indeed the case and, if so, amend the code to parse and join each file.
+    `.replace(/\s+/g, ' '))
+  }
+  const schema = manifest.fileSchema.split(",").map((f) => f.trim());
+  return {
+    schema,
+    inventoryKey: manifest.files[0].key,
+    // Buckets without versioning cannot produce inventories with VersionId
+    versionsExist: schema.includes('VersionId'),
```

Comment:
FWIW, this comment and a few others related to the same `versionsExist` field are, I think, not quite right. As I understand it, whether `VersionId` exists in an inventory is a property of the inventory configuration itself, not the bucket's versioning configuration. That is, I think you can have "all versions" inventory configurations on unversioned buckets; the `VersionId` field will exist but all values will be the empty string. Similarly, you can have "only current" inventory configurations on versioned buckets; there won't be a `VersionId` field but the versions still exist.

Reply:
Thanks! Comment is now updated to reflect this subtlety.